### DISCUSSION TITLE
Package and Organization Name Registrations Should Be Merged

### DISCUSSION THREAD

**Discussion by tolmasky** [npm/feedback#737] [Registry]
I mentioned this on an RFC call a few months ago to positive reception, but was told I'd have to report it to the registry side of things, so I think this is the right place to do it:

Basically, it seems ill-advised that package names and organization names have to be registered separately. It is at best a weird gotcha (you register a package named `whatever`, later realize you want to release plugins or something under the organization-version of the name `@whatever/plugin`, but by then someone has registered the organization), and at worst a security concern: it is very unintuitive that the ownership of the same name can be completely different for a package and an organization. In fact, it seems like an attack vector to simply register un-registered organization names of existing packages in order to pass off scoped packages under the supposed reputation of the same-named package. A particularly nefarious actor could in fact just listen for new package publishes and immediately try to register the associated organization name.

My proposal is the following:

1. For all new package and organization registrations,  packages and organizations effectively occupy the same namespace. That is to say, whether you publish a package named "a" or register an organization named "a", you are immediately granted the permissions to the other.

2. For existing packages and organizations, if there are packages or organizations where the counter-part isn't registered, the current owners are simply granted permissions to those. So, if you currently own package "x", but never bother to register organization "x", you will simply be granted it.

3. Existing package/organization owner mismatches will unfortunately continue to exist as a legacy issue.

(1) is the way I think we wish it always worked, knowing the way people like to organize packages today and the security concerns. (2) "backports" this behavior wherever it is currently applicable, and (3) is simply dealing with the reality that since this wasn't ever enforced in the past there simply exist cases where it will continue to be the case, but we should definitely not continue allowing this.

In other words, "organizations" (scopes) should be a *feature* of a package, not a separate handle open to squatters. The package can be thought of as analogous to the "default export" of a package, and the scope can contain all the "named exports".

**Comment by MylesBorins** [comment]
@tolmasky one complication, scopes are also associated with usernames 😄 

so for this proposal to work we would also have the constraint that:

* a package could not have the name of a username
* if a package exists no username could be made with that package name

I'd be really concerned about the network impact + complexity of enforcing something like this, and to be honest the number of parts of the experience this touches is large enough that it would be very hard for us to prioritize this engineering work on our roadmap at this time.

One idea I've been thinking about, but to be clear I don't think we can do, is no longer allowing global name spaced packages to be published... e.g. all packages must be scoped. This would have the same net effect impact as what your are proposing with far less rules / logic to enforce.

**Comment by tolmasky** [comment]
> @tolmasky one complication, scopes are also associated with usernames 

Wow, honestly I think this makes this change is even more necessary. I focused mainly on the package to organization case above since I think that is the most likely order that people forget to register -- they start with a package and forget to get an organization since they don't think about the future needs of that package (or about someone masquerading as being affiliated with that project...). But with usernames it's the reverse! No one thinks to publish a package under their username just to prevent people from piggy-backing off their reputation. In fact, I just checked, and I hadn't! So thank you for the warning me -- I just published a `tolmasky` package. But to be honest, this username/package situation feels like a much deeper violation, since it hits personally, vs. the organizational situation.

> I'd be really concerned about the network impact + complexity of enforcing something like this, 

I'm not sure I understand the "enforcement" angle here. What I am suggesting is that when you sign up as a new user, you aren't allowed to take a name of an existing user, package, or organization. It's already the case that you can't register an organization with the same name as a user, so it just seems like adding one rule? Just check whether the package name exists and also return an error. The only additional functionality is that whenever you register a username or organization, the package name is then reserved for you. And if you publish a package, the organization is reserved (or created, or whatever) for you. These rules are then retroactively applied where possible to existing username/packages/organizations. I am willing to believe that the codebase is such where these check happen inline in a lot of different places I guess, but I otherwise don't think there's any sort of "active" enforcement that needs to take place.

I also don't know what the term "network impact" is referring to. Perhaps I am misunderstanding the term -- I apologize in advance, but if it refers to "the effect on users", then I certainly don't think this would be surprising or "disruptive" to users in any way. The experience for all existing users is that they suddenly have the rights to certain package names or organizations. No one would even notice that until they went and tried to use those names... and it worked? The only actual "disruption" is that someone trying to publish an as-yet-unpublished package with an existing username or organization-name would receive the same rejection as they currently do if the package name is already taken. There are already cases today where you get that rejection even if the package name doesn't appear to be taken. Most notably, npm will also reject packages that have names too similar to existing packages. npm will also reject package names for packages that appear to be available, but only because they were completely unpublished in the past (I've run into this before). This doesn't really seem to be that different. I'd argue it's less annoying than the similar names thing, which I've run into with valid names that just happen to be close to other package names.

> One idea I've been thinking about, but to be clear I don't think we can do, is no longer allowing global name spaced packages to be published... e.g. all packages must be scoped. This would have the same net effect impact as what your are proposing with far less rules / logic to enforce.

I certainly think that this was probably the way to go from the beginning, and makes the most sense (GitHub certainly doesn't allow top level repositories), but it seems that this would be far more disruptive to the community than the change I proposed. I think people would be legitimately confused if they couldn't publish *any* top level package anymore. Again, to be clear, I'm all for it, but if "network impact" is the consideration this seems like something that would require messaging vs. a change that could honestly probably be deployed completely silently.

> it would be very hard for us to prioritize this engineering work on our roadmap at this time.

I do want to stress that this *is a security issue*. npm *currently* blocks the publishing of packages with names that are too similar to existing packages. If the publishing packages that contain the *exact* name as current packages as the scope is also worrying.

Also, I recommend anyone reading this go publish a package with their username.

**Reply by MylesBorins** [reply]
> I also don't know what the term "network impact" is referring to

The big challenge here is that it doesn't only impact publishing of packages. It means that any time:

* a new package is published
* a new account is created
* a new org is created

We need to validate against the existence of all other potential conflicts and provide an error in a way that is actionable from the developer. For blocking a typo squat it is pretty straight forward... we are comparing packages to pacakges. But when we start blocking pacakges based on usernames or orgs it gets slightly more complex.

It already isn't possible to have a user name or org that match, but if we need to compare the username / org to existing packages things start to get complex. We need to account for fail cases in the CLI, in our website, and in our internal tooling. This isn't impossible to do, but it also is non-trivial from the perspective of implementation and roll out in a way that is not confusing or disruptive to the broader experience.

Another unexpected outcome of rolling out this type of behavior would be an increase in naming disputes. We already have a significant backlog of tickets for folks wanting to get access to packages or scopes that they want the name to and are dormant. Not only would this potentially significantly increase the rate of tickets by doubling the surface area of all dormant packages + usernames, it would also significantly complicate the process of making a decision where transferring a package name is considered a lower risk action than an account.

We already have a fairly robust malware + malicious package reporting process, so in the case of someone doing something malicious with these available package names, we are relatively well covered from a report + response perspective, although I recognize that "quick response" is not as conclusive as "never possible in the first place".

Our roadmap is currently entirely security focused attempting to close the gap on what we have viewed as the most pressing security concern, account takeovers. Our next big chunk of work is planned around securing publishing. Based on what I've outlined above I would imagine that we would need to deprioritize a non-trivial amount for at least a quarter. In my opinion securing accounts and publishing is a higher priority than closing the gap you've identified, even if it is indeed a security concern. If we didn't have existing malware scanning technology, ways to report malicious packages, or a team of folks dedicated to handling these request I might agree that this is a higher risk, but at this time I'm not convinced that it is such a high risk we should deprioritize other roadmap work.

With that said, I don't get to make decisions in a vacuum. If you think this security issue deserves to be prioritized you can always report it [to our hacker one program](https://hackerone.com/github). These reports are triaged by another team that specializes in assessing security risks. If they assess this as high priority we will be required to justify internally what we are prioritizing and why and will be given a deadline to come up with a solution to the problem.

I'm not pointing you in the direction of hacker-one to to pawn you off on "yet another team", but more because I want to be kept honest in this equation and if security is the reason to prioritize this work that is the appropriate channel to get the work prioritized.

I appreciate you taking the time to capture your thoughts here, it would be good to find some time for a short call at some point and we can chat about the broader roadmap and I could get some feedback. We could also brainstorm on a couple other solutions to this space that might be able to close the gap in a less disruptive way
